### PR TITLE
extmod/vfs_posix_file.c: Unix port stdio buffer enhancement

### PR DESCRIPTION
**Pull request by stephanelsmith**
This PR enhances port/unix by adding the buffer attribute to stdio.

Currently, the unix port stdio does not have the buffer attribute as does cpython and other ports, making reading bytes over stdio not possible.  The enhancement will make CLI tools easier to implement, improve porting python tools to micropython, and improved interoperablity between ports, especially when piping raw data samples (IQ, audio, etc...)

Related discussions:
https://github.com/orgs/micropython/discussions/12070
https://github.com/orgs/micropython/discussions/11889

**shared/runtime/sys_stdio_mphal.c** was used as a model for the implementation.  Instead of using **mp_hal_stdin_rx_chr** and **mp_hal_stdout_tx_strn** (which works for raw input but not pipes), I instead created two mp_stream with is_text = false using STDIN_FILE and STDOUT_FILE.  A new stdio buffer stream may then read/write from these streams respectively.

One question I have: I'm unsure of the utilitzation of **MICROPY_PY_SYS_STDIO_BUFFER** as it is used in **sys_stdio_mphal.c**.  I went ahead and made my implementation match though I'm not fully clear of it's intended usage.

I've done some tests, a simple one here piping bytes 0-255 using micropython as well as python.

```
./micropython buffertest.py write | ./micropython buffertest.py read
python buffertest.py write | ./micropython buffertest.py read
./micropython buffertest.py write | python buffertest.py read
```

**buffertest.py**
```
import sys
import asyncio

async def write():
    for x in range(256):
        sys.stdout.buffer.write(x.to_bytes(1,'little'))

async def read():
    while True:
        # r = sys.stdin.read(1)      # today we only have stdin.read which crashes with unicode error at 128
        r = sys.stdin.buffer.read(1) # thumbs up!
        if r != None and len(r)==0:
            break
        print('{:<10} {}'.format(int.from_bytes(r,'big'),r))

async def main():
    if 'read' in sys.argv:
        await read()
    elif 'write' in sys.argv:
        await write()

asyncio.run(main())
```

### CODE REVIEW

File: extmod/vfs_posix_file.c [hunk 1]
```diff
@@ -273,4 +280,49 @@ const mp_obj_vfs_posix_file_t mp_sys_stdin_obj = {{&mp_type_vfs_posix_textio}, S
...
+#if MICROPY_PY_SYS_STDIO_BUFFER
+STATIC const mp_stream_p_t vfs_posix_buffer_stream_p = {
+    .read = vfs_posix_file_read,
+    .write = vfs_posix_file_write,
+    .ioctl = vfs_posix_file_ioctl,
+    .is_text = false,
+};
+MP_DEFINE_CONST_OBJ_TYPE(
+    mp_type_vfs_posix_buffer_inner,
+    MP_QSTR_TextIOWrapper,
```
Comment (dlech) [inline extmod/vfs_posix_file.c:292]: `sys_stdio.mphal.c` uses this name.

```suggestion
    MP_QSTR_FileIO,
```

File: extmod/vfs_posix_file.c [hunk 2]
```diff
@@ -273,4 +280,49 @@ const mp_obj_vfs_posix_file_t mp_sys_stdin_obj = {{&mp_type_vfs_posix_textio}, S
...
+
+STATIC const mp_stream_p_t vfs_posix_buffer_io_stream_p = {
+    .read = stdin_buffer_read,
+    .write = stdout_buffer_write,
+    .ioctl = vfs_posix_file_ioctl,
+    .is_text = false,
+};
+MP_DEFINE_CONST_OBJ_TYPE(
+    mp_type_vfs_posix_buffer,
+    MP_QSTR_TextIOWrapper,
```
Comment (dlech) [inline extmod/vfs_posix_file.c:318]: `sys_stdio.mphal.c` uses this name.

```suggestion
    MP_QSTR_FileIO,
```

File: extmod/vfs_posix_file.c [hunk 3]
```diff
@@ -222,6 +226,9 @@ STATIC mp_uint_t vfs_posix_file_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_
 
 STATIC const mp_rom_map_elem_t vfs_posix_rawfile_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_fileno), MP_ROM_PTR(&vfs_posix_file_fileno_obj) },
+    #if MICROPY_PY_SYS_STDIO_BUFFER
+    { MP_ROM_QSTR(MP_QSTR_buffer), MP_ROM_PTR(&mp_stdio_buffer_obj) },
```
Comment (dlech) [inline extmod/vfs_posix_file.c:230]: Isn't this going to add `buffer` to all file objects, not just `stdin` and `stdout`?

If we do that, then it should provide a buffer object for this specific file instance rather than all objects returning a buffer for stdin/stdout.

**Review by dlech — COMMENTED** [review]
- General question: does it make sense to have `buffer` on all file objects or should only stdio have it?
- This patch only handles stdin and stdout. What about `buffer` for stderr?
- This will also need tests for code coverage.

File: extmod/vfs_posix_file.c (hunk 3 above)
Comment (stephanelsmith) [reply extmod/vfs_posix_file.c:230]: I intended this to just be for stdin and stdout, whereas stdin is read only and stdout is write only.  Similarly, stderr should also be write only.  Right now, it is not doing this, so I need to amend.

More broadly, maybe it's a better solution to have buffer on all file objects.  I will think of this as well.

Regarding code coverage, happy to also do this as well, naive question... but where are they?

Thank you very much for taking a look and providing feedback!

Comment (stephanelsmith) [reply extmod/vfs_posix_file.c:230]: Also fixed the nanbox check and incorporated using FileIO, will work on suggestions....

Comment (dlech) [reply extmod/vfs_posix_file.c:230]: File IO tests are in this directory: https://github.com/micropython/micropython/tree/master/tests/io

If `buffer` is conditionally compiled it probably makes sense to create a new file.

Comment (stephanelsmith) [reply extmod/vfs_posix_file.c]: Hope you are well, have an update on this one:

- I decided to go with adding the `buffer` attribute just on stdin/stdout/stderr files, keeping the scope as was intended for the PR.  For regular files, byte access is already there.

- As `MICROPY_PY_SYS_STDIO_BUFFER` already exists, seems natural to work with it, If `MICROPY_PY_SYS_STDIO_BUFFER` undefined, current bufferless stdio behavior is retained.

For tests, I did two items:

- I updated `file_stdio.py` to include stderr.  I thought it was missing, maybe there's a reason it shouldn't be there?

- I created a new test `file_stdio_buffer.py`.  This one checks to see if buffer attribute is available for stdio items and if so, run the test like `file_stdio.py`, if not, skip.

I wasn't sure if the test should be put in unix folder or io.  For different ports, the stdio tests would not work the same, for example, on esp32, `sys.stdin.fileno()` returns an error as the  `sys_stdio_mphal.c` implementation does not use files, but instead reads from `mp_hal_stdin_rx_chr` and writes to `mp_hal_stdout_tx_strn`.

**Comment by jimmo** [comment]
Thanks for sending the PR @stephanelsmith 

If I understand correctly, the three additional locals dicts are just so that you can add the `buffer` to the attribute to each of stdin/stdout/stderr, but with a different instance for each.

I think this can be solved in a much easier way using a feature which was added since this code was written, which is that you can implement getattr via the `attr` type slot while _also_ providing a locals dict (by having `attr` use the special sentinel return value).

So my proposal would be to add an `attr` handler to `mp_type_vfs_posix_textio`, and the implementation of which compares `self` to the three global instances (`mp_sys_stdin_obj`, etc) and returns the corresponding `mp_sys_stdin_buffer_obj` (which would be of type `mp_type_vfs_posix_fileio`). This could be extended in the future to provide the buffer attribute on regular files.

Something like:

```diff
diff --git a/extmod/vfs_posix_file.c b/extmod/vfs_posix_file.c
index cfc56df99f..e88f2d2a8c 100644
--- a/extmod/vfs_posix_file.c
+++ b/extmod/vfs_posix_file.c
@@ -253,6 +253,7 @@ MP_DEFINE_CONST_OBJ_TYPE(
     locals_dict, &vfs_posix_rawfile_locals_dict
     );
 
+
 STATIC const mp_stream_p_t vfs_posix_textio_stream_p = {
     .read = vfs_posix_file_read,
     .write = vfs_posix_file_write,
@@ -260,12 +261,52 @@ STATIC const mp_stream_p_t vfs_posix_textio_stream_p = {
     .is_text = true,
 };
 
+#if MICROPY_PY_SYS_STDIO_BUFFER
+const mp_obj_vfs_posix_file_t mp_sys_stdin_buffer_obj = {{&mp_type_vfs_posix_fileio}, STDIN_FILENO};
+const mp_obj_vfs_posix_file_t mp_sys_stdout_buffer_obj = {{&mp_type_vfs_posix_fileio}, STDOUT_FILENO};
+const mp_obj_vfs_posix_file_t mp_sys_stderr_buffer_obj = {{&mp_type_vfs_posix_fileio}, STDERR_FILENO};
+
+const mp_obj_vfs_posix_file_t mp_sys_stdin_obj;
+const mp_obj_vfs_posix_file_t mp_sys_stdout_obj;
+const mp_obj_vfs_posix_file_t mp_sys_stderr_obj;
+
+STATIC void vfs_posix_textio_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+    if (dest[0] != MP_OBJ_NULL) {
+        // These objects are read-only.
+        return;
+    }
+
+    if (attr == MP_QSTR_buffer) {
+        // Implement the `buffer` attribute only on std{in,out,err} instances.
+        if (MP_OBJ_TO_PTR(self_in) == &mp_sys_stdin_obj) {
+            dest[0] = MP_OBJ_FROM_PTR(&mp_sys_stdin_buffer_obj);
+            return;
+        }
+        if (MP_OBJ_TO_PTR(self_in) == &mp_sys_stdout_obj) {
+            dest[0] = MP_OBJ_FROM_PTR(&mp_sys_stdout_buffer_obj);
+            return;
+        }
+        if (MP_OBJ_TO_PTR(self_in) == &mp_sys_stderr_obj) {
+            dest[0] = MP_OBJ_FROM_PTR(&mp_sys_stderr_buffer_obj);
+            return;
+        }
+    }
+
+    // Any other attribute - forward to locals dict.
+    dest[1] = MP_OBJ_SENTINEL;
+}
+#define VFS_POSIX_TEXTIO_TYPE_ATTR attr, vfs_posix_textio_attr,
+#else
+#define VFS_POSIX_TEXTIO_TYPE_ATTR
+#endif
+
 MP_DEFINE_CONST_OBJ_TYPE(
     mp_type_vfs_posix_textio,
     MP_QSTR_TextIOWrapper,
     MP_TYPE_FLAG_ITER_IS_STREAM,
     print, vfs_posix_file_print,
     protocol, &vfs_posix_textio_stream_p,
+    VFS_POSIX_TEXTIO_TYPE_ATTR
     locals_dict, &vfs_posix_rawfile_locals_dict
     );
```

**Comment by jimmo** [comment]
@stephanelsmith I ran this past @dpgeorge and he agrees this is the way to go. If you want to update the PR with the suggested diff we can merge it.

As far as writing a test goes, the behavior of text and binary streams differs for reading & writing unicode. See py/stream.c where `is_text` is used. For example, write on a text stream will return number of characters written, on a binary stream.

For example, on CPython:

```
>>> n_text = sys.stdout.write('🚀\n')
🚀
>>> n_binary = sys.stdout.buffer.write('🚀\n'.encode('utf-8'))
🚀
>>> n_text, n_binary
(2, 5)
```

**Comment by stephanelsmith** [comment]
@jimmo Thank you for the review and detailed information! I will have some time to make some updates and re-submit.

File: extmod/vfs_posix_file.c [hunk 4]
```diff
@@ -188,7 +188,7 @@ STATIC mp_uint_t vfs_posix_file_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_
             return 0;
         case MP_STREAM_GET_FILENO:
             return o->fd;
-        #if MICROPY_PY_SELECT
+        #if MICROPY_PY_SELECT && !MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
```
Comment (dlech) [inline extmod/vfs_posix_file.c:191]: This looks unrelated.

File: extmod/vfs_posix_file.c [hunk 5]
```diff
@@ -209,6 +209,15 @@ STATIC mp_uint_t vfs_posix_file_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_
...
                 }
+                if (pfd.revents & POLLERR) {
+                    ret |= MP_STREAM_POLL_ERR;
+                }
+                if (pfd.revents & POLLHUP) {
+                    ret |= MP_STREAM_POLL_HUP;
+                }
+                if (pfd.revents & POLLNVAL) {
+                    ret |= MP_STREAM_POLL_NVAL;
+                }
```
[the remark this answers is not in the document]
Comment (stephanelsmith) [reply extmod/vfs_posix_file.c:220]: yes, this is me using the latest version, this change was made here
https://github.com/micropython/micropython/commit/22106bf2fa574ba947eceb3b9cfa401c67c6a7eb

File: extmod/vfs_posix_file.c (hunk 4 above)
Comment (stephanelsmith) [reply extmod/vfs_posix_file.c:191]: Yes, I'm keeping track of the file as other updates have been made, this change was made in this commit:
https://github.com/micropython/micropython/commit/ef71028f77e4a6fc449c64906ebad0a160f35240

File: extmod/vfs_posix_file.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jimmo) [reply extmod/vfs_posix_file.c:220]: @stephanelsmith The issue is that you need to rebase this PR on the latest master branch, rather than including the same changes in your commit.

(Your commit's parent is cfcce4b5311c7b112d3628392c8012b4c29bdd63 but you need to rebase after 218242d1de781124307be8720a7a9d2f373ca46b)

[the remark this answers is not in the document]
Comment (jimmo) [reply extmod/vfs_posix_file.c]: Thanks!

[the remark this answers is not in the document]
Comment (stephanelsmith) [reply extmod/vfs_posix_file.c]: Rebase done, sorry about that.

**Comment by dlech** [comment]
I think the tests are failing because we don't redirect stderr to stdout when we run CPython at 

https://github.com/micropython/micropython/blob/218242d1de781124307be8720a7a9d2f373ca46b/tests/run-tests.py#L711-L713

like we do when we run MicroPython at

https://github.com/micropython/micropython/blob/218242d1de781124307be8720a7a9d2f373ca46b/tests/run-tests.py#L254-L256

So we probably need a second patch to fix that.

**Comment by stephanelsmith** [comment]
Just getting a moment tonight, I'm still working on the tests.  It's failing on textio test as @jimmo mentioned above where we should be counting characters out for unicode not bytes, matching cpython implementation.  I have an idea here I want to try.

The other question I have is, how do we do a proper test for stdin with pipe?

**Comment by stephanelsmith** [comment]
Alright, think I made some progress tonight
* Updated ```py/stream.c``` ```mp_stream_rw``` function so when writing text, the number output reflects the number of unicode characters, not the number of bytes.  When using buffered stdout/stderr, writes will provide number of bytes.
* Updated ```run-tests.py``` so the CPython expected output includes stderr output.
* Updated ```tests/io/file_stdio_buffer.py``` to test the above conditions, both ascii as well as unicode tests.  All the prints needed to be flushed for consistency between micropython and cpython as the buffering between stdout and stderr could lead to output being inconsistent.

**Comment by stephanelsmith** [comment]
I think I'm looking good here.  Would like to hear what you think.  In review, what as done:
* buffer attr added to stdin/stdout/stderr in `extmod/vfs_posix_file.c`
* `run-tests.py` updated to pipe stderr to output to make stderr testable
* fix made in `py/stream.c` that counts the number of unicode characters instead of bytes when writing
* tests for above added for those in `io/file_stdio2.py`
* addressing a testing corner case for macos, error code 45 added in the `MP_STREAM_FLUSH` allowable error codes for stdout/stderr/stdin in `extmod/vfs_posix_file.c`

**Comment by stephanelsmith** [comment]
Regarding the last item in the bulleted list above...

It's a weird one.  The issue is, on macos, if running micropython from subprocess.check_output, then a stdout.flush() raises error 45.  I don't get this if you run it form the mpy repl on macos.

That said, I can see there's already catching and skipping errors going on with this ```vfs_posix_file.c``` in the ```MP_STREAM_FLUSH``` for stdio.  I've added err 45 as an additional code that can be caught.

In my test, I need to flush line-by-line consistency between cpython and micropython outputs or they come in a bit out of order.  


In case you are curious, here's a minimal test case.  This will run fine on linux, but I can confirm it is crashing on macos with error 45.
```
import sys
import subprocess
import tempfile
with tempfile.NamedTemporaryFile('w') as fp:
    fp.write('''
import sys
sys.stdout.write('hello world')
sys.stdout.flush()
print('')
''')
    fp.flush()
    print('py3')
    o = subprocess.check_output('python3 {}'.format(fp.name).split())
    print(o)
    print('upy')
    o = subprocess.check_output('micropython {}'.format(fp.name).split())
    print(o)
```
```
py3
b'hello world\n'
upy
Traceback (most recent call last):
  File "/var/folders/l4/lf9_l3152yj1xgj6j_47l3tm0000gn/T/tmp3i3uezgj", line 4, in <module>
OSError: 45
```

**Comment by stephanelsmith** [comment]
I may not understand codecov error, but it looks like it's complaining about an unreachable section in `py/stream.c`, the line marked with TODO.  If this is the error, I think there should be no change and mergable.

File: extmod/vfs_posix_file.c [hunk 6]
```diff
@@ -154,11 +154,11 @@ STATIC mp_uint_t vfs_posix_file_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_
         case MP_STREAM_FLUSH: {
             int ret;
             MP_HAL_RETRY_SYSCALL(ret, fsync(o->fd), {
-                if (err == EINVAL
+                if ((err == EINVAL || err == 45)
```
Comment (dlech) [inline extmod/vfs_posix_file.c:157]: I think this should be `ENOTSUP` instead of `45`.

And we could probably also put it inside of `#if __APPLE__` since it is a macos-only issue. (Interestingly, `man fsync` on Mac says that `EINVAL` should be returned when the operation is not supported but apparently that is not actually the case)

File: py/stream.c [hunk 7]
```diff
@@ -75,10 +80,43 @@ mp_uint_t mp_stream_rw(mp_obj_t stream, void *buf_, mp_uint_t size, int *errcode
             return out_sz;
         }
 
+        #if MICROPY_PY_BUILTINS_STR_UNICODE
```
Comment (dlech) [inline py/stream.c:83]: This seems like it is fixing a separate issue and so probably deserves a separate pull request.

File: extmod/vfs_posix_file.c (hunk 6 above)
Comment (stephanelsmith) [reply extmod/vfs_posix_file.c:157]: Done!  I agree.  I looked into it as well.  I added the macro for mac and where it will target both errors.

File: py/stream.c (hunk 7 above)
Comment (stephanelsmith) [reply py/stream.c]: I've removed unicode specific items will add them back in a connected PR that I'll put together next few days.

It would be good to first have this PR merged, so I can add back the unicode tests for stdio which need to removed until unicode support is updated.

Comment (stephanelsmith) [reply py/stream.c]: PR added cover unicode write count here: https://github.com/micropython/micropython/pull/12227

**Comment by jimmo** [comment]
Thanks @stephanelsmith -- sorry I had no idea when I gave that example that we only implement the count-characters vs count-bytes for _read_ and not for _write_. (I just looked quickly in stream.c to see what `is_text` was being used for). Obviously it's a bit harder to test the read side. Thanks for sending the follow-up PR.

I think ideally there should be three separate commits here as there are three independent changes that are happening:
 - run-tests.py change for stderr
 - fix the apple bug (as far as I can tell, this is unrelated to the buffer change, but just came up from testing on macOS?)
 - the actual change + tests.
 
But don't change anything for now, see what Damien thinks first (and we can split it in the merge anyway).

**Comment by stephanelsmith** [comment]
Would love to get this PR merged.  Anyway I can tip the scales.  I see some PRs go through that are git hub sponsored, happy to promote micropython in both code and sponsorship.  This PR as well:
https://github.com/micropython/micropython/pull/12227

This one in particular, but both are for a personal project for high altitude balloon payload I'm developing in micropython, and would love present it to the HAB team.  This is a bit of an intro to micropython for them as well as a way to co-develop on their normal platforms and embedded concurrently.
https://github.com/stephanelsmith/micro-aprs

**Review by dlech — COMMENTED** [review]
Looks OK to me. But Damian is the ultimate gate-keeper here, so we just have to be patient until he can spare some cycles.

**Comment by dpgeorge** [comment]
Thanks everyone involved here for the work implementing and reviewing.

I have split the commits as suggested by @jimmo.
